File: examples/crm-rails/app/models/invitation.rb
class Invitation < ApplicationRecord
  belongs_to :organization
  belongs_to :invited_by, class_name: "User", optional: true

  before_validation { self.token ||= SecureRandom.urlsafe_base64(32) }
  validates :email, presence: true
  validates :role, inclusion: { in: Membership::ROLES }

  def pending? = accepted_at.nil? && revoked_at.nil?

  # Single-use and email-bound, like the platform it showcases.
  def accept!(user)
    raise ArgumentError, "invitation is not pending" unless pending?
    raise ArgumentError, "invitation was issued to a different email" unless user.email == email.strip.downcase

    membership = organization.memberships.create_with(role: role, status: "active")
                             .find_or_create_by!(user: user)
    membership.update!(role: role, status: "active")
    update!(accepted_at: Time.current)
    membership
  end
end

File: examples/employee-directory-laravel/app/Models/Invitation.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Support\Str;

class Invitation extends Model
{
    protected $guarded = [];

    protected static function booted(): void
    {
        static::creating(function (Invitation $invitation) {
            $invitation->token ??= Str::random(48);
        });
    }

    public function organization()
    {
        return $this->belongsTo(Organization::class);
    }

    public function isPending(): bool
    {
        return $this->accepted_at === null && $this->revoked_at === null;
    }

    /** Single-use and email-bound, mirroring the platform semantics. */
    public function accept(User $user): Membership
    {
        if (! $this->isPending()) {
            throw new \InvalidArgumentException('invitation is not pending');
        }
        if (strtolower($user->email) !== strtolower($this->email)) {
            throw new \InvalidArgumentException('invitation was issued to a different email');
        }

        $membership = Membership::updateOrCreate(
            ['organization_id' => $this->organization_id, 'user_id' => $user->id],
            ['role' => $this->role, 'status' => 'active'],
        );
        $this->forceFill(['accepted_at' => now()])->save();

        return $membership;
    }
}
